File: 2218. Maximum Value of K Coins From Piles/Solution.py
class Solution:
    def maxValueOfCoins(self, piles: List[List[int]], k: int) -> int:
        n = len(piles)
        dp = [[0 for i in range(k + 1)] for j in range(n + 1)]
        for i in range(1, n + 1):
            for j in range(k + 1):
                curr = 0
                for l in range(min(j, len(piles[i - 1])) + 1):
                    if l > 0: curr += piles[i - 1][l - 1]
                    dp[i][j] = max(dp[i][j], dp[i - 1][j - l] + curr)
        return dp[n][k]

File: 2218. Maximum Value of K Coins From Piles/Solution.java
class Solution {
    public int maxValueOfCoins(List<List<Integer>> piles, int k) {
        int n = piles.size();
        int[][] dp = new int[n + 1][k + 1];
        for (int i = 1; i <= n; i++) {
            for (int j = 0; j <= k; j++) {
                int curr = 0;
                for (int l = 0; l <= Math.min(piles.get(i - 1).size() , j); l++) {
                    if (l > 0) curr += piles.get(i - 1).get(l - 1);
                    dp[i][j] = Math.max(dp[i][j], dp[i - 1][j - l] + curr);
                }
            }
        }
        return dp[n][k];
    }
}